Add option to debug menu allowing exiting to main menu without saving

LazyWizard opened the pull request:
This PR adds a keybinding and escape menu option that let you quit directly to the main menu without saving. The menu option and keybinding are nonfunctional if the newly-added "Allow quitting without saving" world option isn't enabled.

At the moment this setting can be toggled freely. Ideally it would only be visible at world creation, but I'm new to the codebase and unsure how to go about doing that.

codemime commented:
1. Why a world option?
2. The option should appear only in debug mode. Otherwise it ruins the permadeath concept (which is essential for roguelikes IMHO).

codemime commented on src/game.cpp:
```diff
@@ -3283,6 +3284,17 @@ bool game::handle_action()
             break;
 
+        case ACTION_QUIT_NOSAVE:
+            if (!get_world_option<bool>( "QUIT_NOSAVE" )) {
```
Here and below: code style.

Coolthulhu commented:
Could be just a debug mode bind that doesn't work without debug mode.

keyspace commented:
> On my rig it takes ~30 seconds to alt+F4 out of the game and get back in

Guess this _could_ be useful for "tiles" debugging that take long to load, but IMO this shouldn't be in release builds, even as an option.

> enabled for ... new players who are still learning the ropes and want to try something without ruining their save

Failing horribly is a powerful learning experience. Improvising after failure is the most engaging part. Giving players a designated way to bypass this is prolonging their "still learning the ropes" period.

There are other ways to "back out" anyway - ALT+F4, CTRL+C, killing the process in general, or backing up a save.

kevingranade commented:
Agreed that this would be fine to stick in the debug menu.

Another issue with making it a "regular" option is it wouldn't generally leave the saves in the state the player expected. Incremental map saves happen periodically no matter what game options are selected, so the only thing this will actualy prevent are saves to the player and world save files.  In case it's not obvious, adjusting that save logic in order to allow rolling back the entire state of the game to some previous point isn't something we'd be interested in.

LazyWizard commented:
I pushed a commit moving it to the debug menu. Regarding incremental map saving: what effect would this have? I'm guessing dropped item/vehicle duplication? Would a warning about potential world state weirdness be warranted?

kevingranade commented:
> what effect would this have with quitting directly? I'm guessing potential dropped item/vehicle duplication?

That's the most likely and obvious effect.  I suspect there are a number of other issues that can also occur.

> Would a warning about world state weirdness be warranted?

Maybe, it should be pretty clear to anyone that force-quitting an appliction is not a normal way to exit and you might cause problems.  If you mean whether the command in the debug menu should have a warning, go ahead.

DangerNoodle commented:
I am unsure if it is really worth the effort of preventing the player from savescumming. If they want to do so, let them. Adding this will simply require them to either force-close through task manager, or back up the saves in compressed files.

This is needless work that serves no practical purpose, unless accidentally hitting the close button is a concern. I do not think that is the primary motivation here.

kevingranade commented:
On Dec 29, 2016 7:40 PM, "DangerNoodle" <notifications@github.com> wrote:

I am unsure if it is really worth the effort of preventing the player from
savescumming. If they want to do so, let them. Adding this will simply
require them to either force-close through task manager, or back up the
saves in compressed files.

The project policy is to neither enable nor prevent savescumming. Adding a
non-debug exit would be an enabling feature.

This is needless work...

I'm not sure what needless work you're referencing, adding it as a debug
menu entry is simpler than adding it as a key bind and option.
Also, putting commands in the right place is never needless.

DangerNoodle commented:
I mean that, last I checked, it is still possible to close the game using the window close button, or the requisite macro in case fullscreen is enabled. I assumed this added option would disable the ability to close the window.

If it does not and only allows changing the in-game "quit via suicide" command to not save the world and kill the character , then I am uncertain if the benefits to said option are substantial.

mutability commented:
Re "testing stuff without ruining the save": `chattr +i`

LazyWizard commented:
Due to some poor wording on my part, I'm seeing confusion in these comments about what this PR does. It simply adds an option to the debug menu allowing you to drop back to the main menu without saving first. I renamed the PR to hopefully clear things up, and added one final commit that clarifies things in-game and adds a warning before using.

At this point this PR is ready to be accepted or declined, so I'm removing the [CR] tag.

DangerNoodle commented:
That clears things up, yes. Apologies for the mis-understanding.
